Index/Performance Analsysis changes to DataLocks populate task

IainStevenson opened the pull request:
Potential performance benefits for Submissions and Period end processing.

Commit:
Modified DataLock remote copy to join on UKPRN to reduce Cross Server Nested Loop operations
Copying All Datalocks For UKPRNS
Deleting Non Latest Datalocks locally
Added Significant logging capabilities
Monitoring To [TestStack].[Logs]

Commit:
DPP-939

Commit:
Final Cleanup
Missing Comment from previous version: Now implements NOLOCK on remote read
Dev Tested

IainStevenson commented:
please note this is related to DPP-939 NOT DCM-939 as I named the branch...

MichaelYoung1981 commented on src/SharedPipelineComponents/DataLockEvents/DeployDataLock/sql/dml/02 datalock.populate.reference.history.sql:
```diff
@@ -1,53 +1,263 @@
+SELECT NEWID(),	@logsource,	0,	GETDATE(),	@operation + '- Started: [' + @timestamp + '] Duration: [' + cast(@duration AS VARCHAR(255)) + ']',	NULL
+/* ----- finish pre-op monitoring*/
+
+IF EXISTS (
```
Shouldn't this be an IF NOT EXISTS?  Actually, as you've dropped it earlier in the script is the check necessary?

IainStevenson replied:
Yep, good catch. Last minute changes always are a bad idea! I'll fix and push again. The redundant check is to not rely on assumptions and also for a break in testing.

MichaelYoung1981 commented on src/SharedPipelineComponents/DataLockEvents/DeployDataLock/sql/dml/02 datalock.populate.reference.history.sql:
```diff
@@ -1,53 +1,263 @@
+	[UKPRN]
+	)
+SELECT [UKPRN]
+FROM [DataLock].[vw_Providers]
```
Does this temp table add any value?  Couldn't we just join directly to the view?

IainStevenson replied:
Yes, a further refactor could remove this.

MichaelYoung1981 commented on src/SharedPipelineComponents/DataLockEvents/DeployDataLock/sql/dml/02 datalock.populate.reference.history.sql:
```diff
@@ -1,53 +1,263 @@
+FROM 
+	[Reference].[DataLockEvents] dle
+INNER JOIN 
+	@ProvidersToProcess ptp
```
Is this join necessary?  The query above only pulls events for the providers that have been identified so there wouldn't be a case where there is no record to join to?

IainStevenson replied:
Good point. Its left over from when I pulled over ALL of the data locks. I will see if it can be removed. I think its probably of minimal physical benefit in elapsed time but from a maintenance perspective is much clearer not to. The query used to take about 7-8 seconds, as this stands it takes about 750 milliseconds. I just tested removing it as suggested and its down to 450 ms.... :-)

chrisfoster76 commented:
@IainStevenson this may be a silly question, but are we certain that [TestStack].[Logs] exists in production?

Commit:
Changes following comments:
Convert to use correct logging table
Changed queries to remove step & reduce trasnfer of remote nodes of status value 3
